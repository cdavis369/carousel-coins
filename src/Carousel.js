import { useState } from "react";
import "./Carousel.css";
import Card from "./Card";


/** Carousel: displays images and arrows to navigate through them
 * 
 * Props:
 * - photos: array of {src, caption} objects
 * - title: string describing the collection of images
 * 
 * State:
 * - currCardIdx: integer for current card index
 * 
 * App --> Carousel --> Card
 */
 function Carousel({ photos, title }) {
  const [currCardIdx, setCurrCardIdx] = useState(0);

  const currCard = photos[currCardIdx];
  const total = photos.length;

  //Increments currCardIdx state by 1
  function goForward() {
    setCurrCardIdx(currCardIdx + 1);
  }

  function goBackward() {
    setCurrCardIdx(currCardIdx - 1);
  }


  return (
    <div className="Carousel" data-testid="carousel-container">
      <h1>{title}</h1>
      <div className="Carousel-main">
        {currCardIdx > 0 ? 
          <i
            className="bi bi-arrow-left-circle"
            data-testid="arrow-left"
            onClick={goBackward}
          />
        :
          <i
            className="bi bi-arrow-left-circle"
            data-testid="arrow-left"
            onClick={goBackward}
            hidden
          />
        }
 
        <Card
          caption={currCard.caption}
          src={currCard.src}
          currNum={currCardIdx + 1}
          totalNum={total}
        />
        {currCardIdx < photos.length - 1 ? 
          <i
            className="bi bi-arrow-right-circle"
            data-testid="arrow-right"
            onClick={goForward}
          />
        :
          <i
            className="bi bi-arrow-right-circle"
            data-testid="arrow-right"
            onClick={goForward}
            hidden
          />
        }
      </div>
    </div>
  );
}

export default Carousel;
