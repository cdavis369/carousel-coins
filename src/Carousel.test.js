import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Carousel";
import TEST_IMAGES from "./_testCommon.js";

it("works when you click on the right arrow", function() {
  const { container } = render(
    <Carousel
      photos={TEST_IMAGES}
      title="images for testing"
    />
  );
  // expect the first image to show, but not the second
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).not.toBeInTheDocument();

  // move forward in the carousel
  const rightArrow = container.querySelector(".bi-arrow-right-circle");
  fireEvent.click(rightArrow);

  // expect the second image to show, but not the first
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).not.toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).toBeInTheDocument();
});

// Smoke test.
test('renders Carousel component without crashing', () => {
  const photos = [
    { caption: 'caption test 1', src: TEST_IMAGES[0].src },
    { caption: 'caption test 2', src: TEST_IMAGES[1].src },
  ];
  render(<Carousel photos={TEST_IMAGES} title="testing title" />);
})

// Snapshot test.
test('matches snapshot', () => {
  const { asFragment } = render(<Carousel photos={TEST_IMAGES} title="testing title" />);
  expect(asFragment()).toMatchSnapshot();
});

// Left arrow test.
test('clicking left arrow moves to previous image', () => {
  const { getByTestId } = render(<Carousel photos={TEST_IMAGES} title="testing title" />);
  expect(getByTestId('carousel-container')).toHaveTextContent('testing image 1');

  fireEvent.click(getByTestId('arrow-right'));
  expect(getByTestId('carousel-container')).toHaveTextContent('testing image 2');

  fireEvent.click(getByTestId('arrow-left'));
  expect(getByTestId('carousel-container')).toHaveTextContent('testing image 1');
});

// Left arrow hidden test
test('left arrow is hidden on carousel render', () => {
  const {getByTestId} = render(<Carousel photos={TEST_IMAGES} title="testing title" />);
  expect(getByTestId('arrow-left')).toHaveAttribute('hidden')
});

// Right arrow hidden
test('right arrow is hidden when carousel on last image', () => {
  const {getByTestId} = render(<Carousel photos={TEST_IMAGES} title="testing title" />);
  fireEvent.click(getByTestId('arrow-right'));
  fireEvent.click(getByTestId('arrow-right'));
  expect(getByTestId('arrow-right')).toHaveAttribute('hidden')
});