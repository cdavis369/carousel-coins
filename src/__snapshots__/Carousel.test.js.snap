// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
    data-testid="carousel-container"
  >
    <h1>
      testing title
    </h1>
    <div
      class="Carousel-main"
    >
      <i
        class="bi bi-arrow-left-circle"
        data-testid="arrow-left"
        hidden=""
      />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          testing image 1
        </h4>
        <img
          alt="testing image 1"
          class="Card-image"
          src="test1.com"
        />
        <small
          class="Card-small"
        >
          Image 1 of 3.
        </small>
      </div>
      <i
        class="bi bi-arrow-right-circle"
        data-testid="arrow-right"
      />
    </div>
  </div>
</DocumentFragment>
`;
